### ISSUE TITLE
Remove freezer special case

### ISSUE THREAD

**Issue by dwightguth** [kframework/k-legacy#465]
We decided that freezers will just become special klabels. For example, the context:

```
context HOLE + _:Exp
```

Will become the heating and cooling rules:
rule E:Exp + Fresh:Exp => E ~> 'HOLE+_(Fresh)
rule R:KResult ~> 'HOLE+_(Fresh) => R + Fresh

This needs to be implemented by modifying the ContextsToHeating transformer.

**Comment by traiansf** [comment]
This might prove problematic to implement in its full generality.

First, we have discussed at some point (and it's almost implemented) the possibility of having nested freezers (a good example for this would be the ability to encode the call-by-need lambda calculus).  to do so, we need to be able to match against a freezed term in which case preserving the structure would be helpful.

Second, perhaps less important, some contexts might be more involved, so finding a nice representation for them could be difficult.

Of course, both are solvable;  however, I would advocate for maintaining the structure in the ast representation (using a special label freezer and as special token HOLE), and only flatten them in the backends (if we decide this is important)

**Comment by dwightguth** [comment]
Hmm, alright. It's worth considering. However, the purpose of this change
is to remove the instance of KInjectedLabel which is deprecated. There is
another solution we could pursue which is equally allowable. Namely:

syntax KLabel ::= "#freezer"
syntax KToken :=  HOLE

This still requires changes to the code base in order to implement, and I
am agnostic as to which route we pursue. I merely deferred to Grigore's
proposed solution when writing this, that's all.

**Comment by grosu** [comment]
As we know, using _+_(HOLE,3), etc., can lead to inconsistencies if we allow heating cooling to happen anywhere.  On the other hand, if we only allow heating/cooling at the top of the k cell, then we should be OK.  However, we still have to be careful to block rewriting when things with holes in them reach the top of k.  For example, we should not allow heating <k> _+_(HOLE,x) ...</k>.  But we can do this later, when it becomes a problem.

In short, I think it is OK to just add HOLE as a KToken constant and thus deal with freezers elegantly  by just using the already existing labels plus the HOLE.  Dwight, why use `#freezer` as a label?  To inhibit heating/cooling underneath?  Then you may want to call it `#frozen` instead.  This could be a good idea, but I would start with the simpler approach, where we just add the HOLE KToken.

**Comment by traiansf** [comment]
Actually, if we want to eventually support nested freezers, we need the
hole to be parameterized by the nesting level, too, as it is now in the
Java KAST.

But, to be clear,  I only see this as a representation means, including the
#freezer label to delimit the freezer;  In the code which actually runs, We
can rename freezers to freshly generated unique labels (as long as krun
knows to unparse them).

2014-05-17 0:28 GMT+03:00 Grigore Rosu notifications@github.com:
